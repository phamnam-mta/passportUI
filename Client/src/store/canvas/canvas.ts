import { createSlice } from "@reduxjs/toolkit";
import { clearCurrentDocument, setCurrentDocument, setCurrentPage } from "store/documents/documents";
import { IRawDocument } from "store/documents/documentsTypes";

export interface ICanvas {
    imageUrl: string;
    width: number;
    height: number;
    angle: number;
}

export enum VisibleAnalyzedElementEnum {
    KeyValuePairs = "KeyValuePairs",
    Entities = "Entities",
    PagedLabelResult = "PagedLabelResult",
    Lines = "Lines",
    Words = "Words",
    Paragraphs = "Paragraphs",
    SelectionMarks = "SelectionMarks",
    Tables = "Tables",
}

export type VisibleAnalyzedElement = {
    [VisibleAnalyzedElementEnum.KeyValuePairs]?: boolean;
    [VisibleAnalyzedElementEnum.Entities]?: boolean;
    [VisibleAnalyzedElementEnum.PagedLabelResult]?: boolean;
    [VisibleAnalyzedElementEnum.Lines]?: boolean;
    [VisibleAnalyzedElementEnum.Words]: boolean;
    [VisibleAnalyzedElementEnum.Paragraphs]?: boolean;
    [VisibleAnalyzedElementEnum.Tables]?: boolean;
    [VisibleAnalyzedElementEnum.SelectionMarks]?: boolean;
};

export interface loadDocumentPagePayload {
    document: IRawDocument;
    pageNumber: number;
}

export type CanvasState = {
    canvas: ICanvas;
    visibleAnalyzedElement: VisibleAnalyzedElement;
    hoveredBoundingBoxIds: string[];
    hoveredLabelName: string;
    documentSelectIndex: number;
    shouldResizeImageMap: boolean;
};

export const initialState: CanvasState = {
    canvas: { imageUrl: "", width: 0, height: 0, angle: 0 },
    documentSelectIndex: 0,
    visibleAnalyzedElement: { [VisibleAnalyzedElementEnum.Words]: true },
    hoveredBoundingBoxIds: [],
    hoveredLabelName: "",
    shouldResizeImageMap: false,
};

const canvasSlice = createSlice({
    name: "canvas",
    initialState,
    reducers: {
        setAngle(state, action) {
            state.canvas.angle = action.payload;
        },
        setVisibleAnalyzedElement(state, action) {
            const { element, value } = action.payload;
            state.visibleAnalyzedElement[element] = value;
        },
        setHoveredBoundingBoxIds(state, action) {
            state.hoveredBoundingBoxIds = action.payload;
        },
        setHoveredLabelName(state, action) {
            state.hoveredLabelName = action.payload;
        },
        setDocumentSelectIndex(state, action) {
            state.documentSelectIndex = action.payload;
        },
        setShouldResizeImageMap(state, action) {
            state.shouldResizeImageMap = action.payload;
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(clearCurrentDocument, (state) => {
                state.canvas = { imageUrl: "", width: 0, height: 0, angle: 0 };
                state.hoveredBoundingBoxIds = [];
                state.hoveredLabelName = "";
                state.documentSelectIndex = 0;
            })
            .addCase(setCurrentDocument.fulfilled, (state, action) => {
                const { imageUrl, width, height, angle } = action.payload.documentPage;
                state.canvas = { imageUrl, width, height, angle };
                state.hoveredBoundingBoxIds = [];
                state.hoveredLabelName = "";
                state.documentSelectIndex = 0;
            })
            .addCase(setCurrentPage.fulfilled, (state, action) => {
                state.canvas = action.payload.documentPage;
                state.hoveredBoundingBoxIds = [];
                state.hoveredLabelName = "";
            });
    },
});

export const {
    setAngle,
    setVisibleAnalyzedElement,
    setHoveredBoundingBoxIds,
    setHoveredLabelName,
    setDocumentSelectIndex,
    setShouldResizeImageMap,
} = canvasSlice.actions;

export const reducer = canvasSlice.reducer;
