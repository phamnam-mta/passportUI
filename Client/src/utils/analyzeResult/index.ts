import { BoundingRegion, Polygon, StudioFieldValue, StudioSpan } from "models/analyzeResult";

export const getPagePolygons = (boundingRegions: BoundingRegion[], page: number): Polygon[] => {
    return boundingRegions
        .filter((boundingRegion) => boundingRegion.pageNumber === page)
        .map((boundingRegion) => boundingRegion.polygon);
};

export const isAnyBoundingRegionInPage = (boundingRegions: BoundingRegion[], page: number): boolean => {
    return boundingRegions.some((boundingRegion) => boundingRegion.pageNumber === page);
};

export const getFieldTypeValue = (field: StudioFieldValue) => {
    const { type } = field;
    const typeString = type.charAt(0).toUpperCase() + type.slice(1);

    // TODO: Update when going public.
    if (type === "currency" && field.valueCurrency && typeof field.valueCurrency !== "string") {
        const { currencySymbol, currency, amount } = field.valueCurrency;
        if (currencySymbol && amount) {
            return `${currencySymbol}${amount}`;
        } else if (currency && amount) {
            return `${currency}${amount}`;
        } else if (amount) {
            return `${amount}`;
        } else {
            return undefined;
        }
    } else if (type === "address" && field.valueAddress) {
        return field.content;
    }
    return field[`value${typeString}`];
};

export interface SpanPosition extends StudioSpan {
    iconName: string;
    pageNumber?: number;
    tableIndex?: number;
    shouldDisplayAsIcon: boolean;
}
