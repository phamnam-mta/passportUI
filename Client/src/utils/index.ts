export const delay = (ms: number) => {
    return new Promise<void>((resolve) => {
        setTimeout(() => {
            resolve();
        }, ms);
    });
};

export const getPixelWidthFromPercent = (widthInPercentage: number): number => {
    const totalWindowWidth = document.documentElement.clientWidth;
    const paneWidthInPixel = (totalWindowWidth / 100) * widthInPercentage;
    return paneWidthInPixel;
};

export const loadCanvasToBlob = (canvas: HTMLCanvasElement): Promise<Blob> => {
    return new Promise((resolve, reject) => {
        canvas.toBlob((blob) => {
            if (blob) {
                resolve(blob);
            } else {
                reject("Failed to create blob for canvas.");
            }
        });
    });
};

export const loadUrlToArrayBuffer = async (url: string): Promise<ArrayBuffer> => {
    return await fetch(url).then((r) => r.arrayBuffer());
};

export const debounce = (func, timeout = 300) => {
    return (...args) => {
        return setTimeout(() => {
            func.apply(this, args);
        }, timeout);
    };
};
