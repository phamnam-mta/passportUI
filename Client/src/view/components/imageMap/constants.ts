export const LAYER_NAME = "name";
export const IMAGE_LAYER_NAME = "imageLayer";
export const TEXT_VECTOR_LAYER_NAME = "textVectorLayer";
export const POD_VECTOR_LAYER_NAME = "podVectorLayer";
export const TABLE_BORDER_VECTOR_LAYER_NAME = "tableBorderVectorLayer";
export const TABLE_ICON_VECTOR_LAYER_NAME = "tableIconVectorLayer";
export const TABLE_ICON_BORDER_VECTOR_LAYER_NAME = "tableIconBorderVectorLayer";
export const CHECKBOX_VECTOR_LAYER_NAME = "checkboxBorderVectorLayer";
export const LABEL_VECTOR_LAYER_NAME = "labelledVectorLayer";
export const DRAWN_REGION_LABEL_VECTOR_LAYER_NAME = "drawnRegionLabelledVectorLayer";
export const DRAWN_REGION_VECTOR_LAYER_NAME = "drawnRegionVectorLayer";

export const SELECTED_PROPERTY = "selected";
export const SELECTION_MARK_STATE = "selectionMarkState";
export const HIGHLIGHTED_PROPERTY = "highlighted";
export const FIELD_PROPERTY = "field";
export const COLOR_PROPERTY = "color";
export const BACKGROUND_COLOR_PROPERTY = "backgroundColor";
export const DASHED_PROPERTY = "dashed";
export const DISABLEHIGHLIGHT_PROPERTY = "disableHighlight";
