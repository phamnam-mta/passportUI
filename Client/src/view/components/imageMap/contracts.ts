/**
 * @name - Feature Category
 * @description - Defines types of feature
 * @member Checkbox - Checkbox
 * @member DrawnRegion - User drawn region
 * @member Label - User label
 * @member Text - OCR text
 */
export enum FeatureCategory {
    Checkbox = "checkbox",
    DrawnRegion = "region",
    Label = "label",
    Text = "text",
}

/**
 * @name - Region Type
 * @description - Defines the region type within the asset metadata
 * @member Point - Specifies a vertex
 * @member Polygon - Specifies a region as a multi-point polygon
 * @member Polyline - Specifies a region as a multi-point line
 * @member Square - Specifies a region as a square
 * @member Rectangle - Specifies a region as a rectangle
 */
export enum RegionType {
    Point = "POINT",
    Polygon = "POLYGON",
    Polyline = "POLYLINE",
    Rectangle = "RECTANGLE",
    Square = "SQUARE",
}

/**
 * @name - Region
 * @description - Defines a region within an asset
 * @member id - Unique identifier for this region
 * @member type - Defines the type of region
 * @member category - Defines the feature category
 * @member tags - Defines a list of tags applied to a region
 * @member points - Defines a list of points that define a region
 * @member boundingBox - Defines a list of points that forms a bounding box
 * @member value - Defines a normalized value for this region
 * @member pageNumber - Defines a page number for this region
 * @member isTableRegion - Defines a flag determine if the region belongs to a table
 * @member changed - Defines a flag if this region had been changed
 */
export interface IRegion {
    id: string;
    type: RegionType;
    category: FeatureCategory;
    tags: string[];
    points?: IPoint[];
    boundingBox?: IBoundingBox;
    value?: string;
    pageNumber: number;
    isTableRegion?: boolean;
    changed?: boolean;
}

/**
 * @name - Bounding Box
 * @description - Defines the tag usage within a bounding box region
 * @member left - Defines the left x boundary for the start of the bounding box
 * @member top - Defines the top y boundary for the start of the boudning box
 * @member width - Defines the width of the bounding box
 * @member height - Defines the height of the bounding box
 */
export interface IBoundingBox {
    left: number;
    top: number;
    width: number;
    height: number;
}

/**
 * @name - Point
 * @description - Defines a point / coordinate within a region
 * @member x - The x value relative to the asset
 * @member y - The y value relative to the asset
 */
export interface IPoint {
    x: number;
    y: number;
}
