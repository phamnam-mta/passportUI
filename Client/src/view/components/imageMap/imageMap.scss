@import "~@fluentui/react/dist/sass/References";

.map-wrapper {
    display: flex;
    flex-grow: 1;
    background-color: $ms-color-gray20;

    #map {
        width: 100%;
        height: 100%;
    }
}
