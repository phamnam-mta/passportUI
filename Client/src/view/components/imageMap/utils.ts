import { Feature } from "ol";
import { Extent } from "ol/extent";
import Polygon from "ol/geom/Polygon";
import ImageLayer from "ol/layer/Image";
import Layer from "ol/layer/Layer";
import VectorLayer from "ol/layer/Vector";
import Projection from "ol/proj/Projection";
import ImageSource from "ol/source/ImageStatic";
import VectorSource from "ol/source/Vector";
import { LAYER_NAME } from "./constants";
import { IRegion } from "./contracts";

export type Dimension = {
    width: number;
    height: number;
};

export const degreeToRadians = (degree: number) => {
    // convert degree to radians
    return (degree * Math.PI * 2) / 360;
};

export const makeLayerFilter = (layerName) => {
    return {
        layerFilter: (layer: Layer) => layer.get(LAYER_NAME) === layerName,
    };
};

export const makeVectorLayer = (layerName, options = {}): Layer => {
    return new VectorLayer({
        [LAYER_NAME]: layerName,
        source: new VectorSource(),
        ...options,
    } as any);
};

export const makeImageLayer = (layerName, uri, projection, imageExtent, options = {}): Layer => {
    return new ImageLayer({
        [LAYER_NAME]: layerName,
        source: makeImageSource(uri, projection, imageExtent),
        ...options,
    } as any);
};

export const makeImageSource = (url: string, projection: Projection, imageExtent: Extent) => {
    return new ImageSource({ url, projection, imageExtent });
};

export const convertRegionToFeature = (region: IRegion, imageExtent: Extent, isOcrProposal: boolean = false) => {
    const coordinates: Array<[number, number]> = [];
    const boundingBox = region.id.split(",").map(parseFloat);
    const imageWidth = imageExtent[2] - imageExtent[0];
    const imageHeight = imageExtent[3] - imageExtent[1];
    for (let i = 0; i < boundingBox.length; i += 2) {
        coordinates.push([Math.round(boundingBox[i] * imageWidth), Math.round((1 - boundingBox[i + 1]) * imageHeight)]);
    }
    const feature = new Feature({
        geometry: new Polygon([coordinates]),
    });
    feature.setProperties({
        id: region.id,
        text: region.value,
        highlighted: false,
        isOcrProposal,
    });
    feature.setId(region.id);

    return feature;
};

export const createRegionIdFromPolygon = (polygon: number[], page: number): string => {
    return polygon.join(",") + ":" + page;
};

export const convertToImageMapCoordinates = (polygon: number[], canvasSize: Dimension, documentSize: Dimension) => {
    const { width: canvasWidth, height: canvasHeight } = canvasSize;
    const { width: ocrWidth, height: ocrHeight } = documentSize;

    const coordinates: number[][] = [];
    for (let i = 0; i < polygon.length; i += 2) {
        coordinates.push([(polygon[i] / ocrWidth) * canvasWidth, (1 - polygon[i + 1] / ocrHeight) * canvasHeight]);
    }
    return coordinates;
};

export const getBoundingBoxFromFeatureId = (id) => id.split(",").map(parseFloat);
