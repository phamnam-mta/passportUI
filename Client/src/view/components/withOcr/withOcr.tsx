import * as React from "react";
import { Feature } from "ol";
import { Extent } from "ol/extent";
import Polygon from "ol/geom/Polygon";
import { compose } from "redux";
import { connect } from "react-redux";
import { ApplicationState } from "store";
import { setHoveredBoundingBoxIds, VisibleAnalyzedElementEnum } from "store/canvas/canvas";
import { AnalyzeResultAdapterFactory } from "adapters/analyzeResultAdapter";
import { ImageMap } from "../imageMap/imageMap";
import { createRegionIdFromPolygon, convertToImageMapCoordinates } from "../imageMap/utils";
import { FeatureCategory } from "../imageMap/contracts";
import { FIELD_PROPERTY } from "../imageMap/constants";
import { checkboxStyler } from "utils/styler";

export const withOcr = (ImageMapComponent) => {
    class WithOcr extends React.PureComponent<any> {
        private imageMap: ImageMap | null = null;

        public async componentDidUpdate(prevProps: any) {
            const { currentDocument, predictions, visibleAnalyzedElement } = this.props;
            if (prevProps.currentDocument !== currentDocument && currentDocument) {
                this.clearLayout();
                if (
                    predictions[currentDocument.name] &&
                    predictions[currentDocument.name].analyzeResponse.analyzeResult
                ) {
                    this.drawLayout(currentDocument.currentPage);
                }
            }

            if (prevProps.predictions !== predictions && currentDocument) {
                if (
                    predictions[currentDocument.name] &&
                    predictions[currentDocument.name].analyzeResponse.analyzeResult
                ) {
                    this.clearLayout();
                    this.drawLayout(currentDocument.currentPage);
                }
            }

            if (prevProps.visibleAnalyzedElement !== visibleAnalyzedElement && currentDocument) {
                if (
                    predictions[currentDocument.name] &&
                    predictions[currentDocument.name].analyzeResponse.analyzeResult
                ) {
                    this.clearLayout();
                    this.drawLayout(currentDocument.currentPage);
                }
            }
        }

        private setImageMap = (ref: ImageMap) => {
            this.imageMap = ref;
            if (this.props.setImageMap) {
                this.props.setImageMap(ref);
            }
        };

        private getAnalyzeResult() {
            const { currentDocument, predictions } = this.props;
            return predictions[currentDocument.name].analyzeResponse.analyzeResult;
        }

        private clearLayout() {
            this.imageMap?.removeAllTextFeatures();
            this.imageMap?.removeAllCheckboxFeatures();
        }

        private drawLayout = (targetPage: number) => {
            const analyzeResultAdapter = AnalyzeResultAdapterFactory.create(this.getAnalyzeResult());
            const documentPage = analyzeResultAdapter.getDocumentPage(targetPage);
            const imageExtent = this.imageMap?.getImageExtent() as Extent;
            const textFeatures: Feature[] = [];
            const lineFeatures: Feature[] = [];
            const selectionMarkFeatures: Feature[] = [];

            if (!documentPage) {
                return;
            }

            const { pageNumber, width, height, words, selectionMarks, lines } = documentPage;
            const ocrExtent: Extent = [0, 0, width, height];

            if (this.props.visibleAnalyzedElement[VisibleAnalyzedElementEnum.Lines]) {
                const feature = this.createPrebuiltLineFeatures(lines, imageExtent, ocrExtent, pageNumber);
                lineFeatures.push(...feature);
            }

            if (this.props.visibleAnalyzedElement[VisibleAnalyzedElementEnum.Words]) {
                words.forEach((word) => {
                    const { content, polygon } = word;
                    textFeatures.push(
                        this.createFeature(content, polygon, imageExtent, ocrExtent, pageNumber, FeatureCategory.Text)
                    );
                });
            }

            (selectionMarks || []).forEach((selectionMark) => {
                const { state, polygon } = selectionMark;
                selectionMarkFeatures.push(
                    this.createFeature(
                        state,
                        polygon,
                        imageExtent,
                        ocrExtent,
                        pageNumber,
                        FeatureCategory.Checkbox,
                        selectionMark
                    )
                );
            });

            if (textFeatures.length > 0) {
                this.imageMap?.addFeatures(textFeatures);
            }

            if (lineFeatures.length > 0) {
                this.imageMap?.addFeatures(lineFeatures);
            }

            if (selectionMarkFeatures.length > 0) {
                this.imageMap?.addCheckboxFeatures(selectionMarkFeatures);
            }
        };

        private createPrebuiltLineFeatures = (lines: any, imageExtent: Extent, ocrExtent: Extent, page: number) => {
            if (!lines) {
                return [];
            }
            const features: Feature[] = [];
            const canvasSize = {
                width: imageExtent[2] - imageExtent[0],
                height: imageExtent[3] - imageExtent[1],
            };
            const documentSize = {
                width: ocrExtent[2] - ocrExtent[0],
                height: ocrExtent[3] - ocrExtent[1],
            };
            lines.forEach((lineItem) => {
                const featureId = createRegionIdFromPolygon(lineItem.polygon, page);
                const coordinates: number[][] = convertToImageMapCoordinates(
                    lineItem.polygon,
                    canvasSize,
                    documentSize
                );
                const feature = new Feature({
                    geometry: new Polygon([coordinates]),
                    id: featureId,
                    [FIELD_PROPERTY]: lineItem,
                });
                feature.setId(featureId);
                features.push(feature);
            });

            return features;
        };

        private createFeature = (
            text: string,
            polygon: number[],
            imageExtent: Extent,
            ocrExtent: Extent,
            page: number,
            category: FeatureCategory,
            fieldItem?: any
        ) => {
            const coordinates: any[] = [];
            const polygonPoints: number[] = [];

            // An array of numbers representing an extent: [minx, miny, maxx, maxy]
            const imageWidth = imageExtent[2] - imageExtent[0];
            const imageHeight = imageExtent[3] - imageExtent[1];
            const ocrWidth = ocrExtent[2] - ocrExtent[0];
            const ocrHeight = ocrExtent[3] - ocrExtent[1];

            for (let i = 0; i < polygon.length; i += 2) {
                coordinates.push([
                    Math.round((polygon[i] / ocrWidth) * imageWidth),
                    Math.round((1 - polygon[i + 1] / ocrHeight) * imageHeight),
                ]);
                polygonPoints.push(polygon[i] / ocrWidth);
                polygonPoints.push(polygon[i + 1] / ocrHeight);
            }

            const featureId = createRegionIdFromPolygon(polygonPoints, page);
            const feature = new Feature({
                geometry: new Polygon([coordinates]),
                id: featureId,
                text,
                polygon: polygon,
                highlighted: false,
                isOcrProposal: true,
                category,
                [FIELD_PROPERTY]: fieldItem,
            });
            feature.setId(featureId);

            return feature;
        };

        public render() {
            return (
                <ImageMapComponent
                    {...this.props}
                    setImageMap={this.setImageMap}
                    checkboxFeatureStyler={checkboxStyler}
                />
            );
        }
    }

    return WithOcr;
};

const mapState = (state: ApplicationState) => ({
    visibleAnalyzedElement: state.canvas.visibleAnalyzedElement,
    hoveredBoundingBoxIds: state.canvas.hoveredBoundingBoxIds,
    documentSelectIndex: state.canvas.documentSelectIndex,
});

const mapDispatch = { setHoveredBoundingBoxIds };

const connector = connect(mapState, mapDispatch);

const composedComponent = compose<any>(connector, withOcr);

export default composedComponent;
