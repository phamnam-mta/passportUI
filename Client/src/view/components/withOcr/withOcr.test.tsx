import * as React from "react";
import { shallow } from "enzyme";
import { Feature } from "ol";
import { VisibleAnalyzedElementEnum } from "store/canvas/canvas";
import { mockDocument } from "utils/test";
import { mockLayoutV3_0_3_AnalyzeResponse } from "utils/test/mockAnalyzeData/v3_0_3_mockAnalyzeData";
import { ImageMap } from "../imageMap/imageMap";
import { withOcr } from "./withOcr";

jest.mock("../imageMap/imageMap");

describe("withOcr", () => {
    const WithOcr = withOcr(ImageMap);

    const currentDocument = mockDocument;

    const predictions = {
        [mockDocument.name]: {
            name: mockDocument.name,
            analyzeResponse: mockLayoutV3_0_3_AnalyzeResponse,
        },
    };

    let baseProps;
    let mockImageMap;
    let wrapper;
    let instance;
    beforeEach(() => {
        baseProps = {
            setImageMap: jest.fn(),
            currentDocument: currentDocument,
            predictions: {},
            visibleAnalyzedElement: {
                [VisibleAnalyzedElementEnum.Lines]: false,
                [VisibleAnalyzedElementEnum.Words]: true,
            },
        };

        mockImageMap = {
            addCheckboxFeatures: jest.fn(),
            addFeatures: jest.fn(),
            getImageExtent: () => [0, 0, 850, 1100],
            removeAllCheckboxFeatures: jest.fn(),
            removeAllTextFeatures: jest.fn(),
        };
        wrapper = shallow(<WithOcr {...baseProps} />);
        instance = wrapper.instance() as any;
        instance.setImageMap(mockImageMap);
    });

    describe("Rendering", () => {
        it("should render correctly", () => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe("Event Handling", () => {
        it("should propagate setImageMap", () => {
            expect(instance.imageMap).toBeTruthy();
            expect(baseProps.setImageMap).toBeCalledTimes(1);
        });

        it("should draw layout features, when layout results are set.", () => {
            wrapper.setProps({ predictions });
            expect(mockImageMap.removeAllCheckboxFeatures).toBeCalledTimes(1);
            expect(mockImageMap.removeAllTextFeatures).toBeCalledTimes(1);
            expect(mockImageMap.addCheckboxFeatures).toBeCalledTimes(1);
            expect(mockImageMap.addFeatures).toBeCalledTimes(1);
        });

        it("should re-draw layout feature, when visibleAnalyzedElement is changed ", () => {
            const props = { ...baseProps, predictions };
            const wrapper = shallow(<WithOcr {...props} />);
            const instance = wrapper.instance() as any;
            instance.setImageMap(mockImageMap);

            wrapper.setProps({
                visibleAnalyzedElement: { [VisibleAnalyzedElementEnum.Lines]: true },
            });

            expect(mockImageMap.removeAllCheckboxFeatures).toBeCalledTimes(1);
            expect(mockImageMap.removeAllTextFeatures).toBeCalledTimes(1);
            expect(mockImageMap.addCheckboxFeatures).toBeCalledTimes(1);
            expect(mockImageMap.addFeatures).toBeCalledTimes(1);
        });

        it("should draw layout features, when switch to analyzed document.", () => {
            const notAnalyzedDocument = {
                ...currentDocument,
                name: "fake-document.pdf",
                states: { loadingStatus: "Loaded" },
            };
            const props = { ...baseProps, currentDocument: notAnalyzedDocument, predictions };
            const wrapper = shallow(<WithOcr {...props} />);
            const instance = wrapper.instance() as any;
            instance.setImageMap(mockImageMap);

            wrapper.setProps({ currentDocument });
            expect(mockImageMap.removeAllCheckboxFeatures).toBeCalledTimes(1);
            expect(mockImageMap.removeAllTextFeatures).toBeCalledTimes(1);
            expect(mockImageMap.addCheckboxFeatures).toBeCalledTimes(1);
            expect(mockImageMap.addFeatures).toBeCalledTimes(1);
        });

        it("should clear layout features, when switch to not-analyzed document.", () => {
            const notAnalyzedDocument = {
                ...currentDocument,
                name: "not-analyzed-document.pdf",
                states: { loadingStatus: "Loaded" },
            };
            wrapper.setProps({ currentDocument: notAnalyzedDocument });
            expect(mockImageMap.removeAllCheckboxFeatures).toBeCalledTimes(1);
            expect(mockImageMap.removeAllTextFeatures).toBeCalledTimes(1);
            expect(mockImageMap.addCheckboxFeatures).not.toBeCalled();
            expect(mockImageMap.addFeatures).not.toBeCalled();
        });

        it("should create expected layout features, when layout analyze results are given.", () => {
            wrapper.setProps({ predictions });
            expect(mockImageMap.removeAllCheckboxFeatures).toBeCalledTimes(1);
            expect(mockImageMap.removeAllTextFeatures).toBeCalledTimes(1);
            expect(mockImageMap.addCheckboxFeatures).toBeCalledTimes(1);
            expect(mockImageMap.addFeatures).toBeCalledTimes(1);

            const expectTextCount = 205;
            const addedTextFeatures = mockImageMap.addFeatures.mock.calls[0][0];
            expect(addedTextFeatures).toHaveLength(expectTextCount);
            addedTextFeatures.forEach((addedFeature) => {
                expect(addedFeature).toBeInstanceOf(Feature);
            });

            const expectCheckboxCount = 25;
            const addedCheckboxFeatures = mockImageMap.addCheckboxFeatures.mock.calls[0][0];
            expect(addedCheckboxFeatures).toHaveLength(expectCheckboxCount);
            addedCheckboxFeatures.forEach((checkbox) => {
                expect(checkbox).toBeInstanceOf(Feature);
            });
        });
    });
});
