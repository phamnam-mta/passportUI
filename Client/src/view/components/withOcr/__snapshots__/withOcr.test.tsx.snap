// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withOcr Rendering should render correctly 1`] = `
<ImageMap
  checkboxFeatureStyler={[Function]}
  currentDocument={
    Object {
      "currentPage": 1,
      "name": "test.pdf",
      "numPages": 10,
      "states": Object {
        "loadingStatus": "Loaded",
      },
      "thumbnail": "",
      "type": "application/pdf",
      "url": "<url>",
    }
  }
  predictions={Object {}}
  setImageMap={[Function]}
  visibleAnalyzedElement={
    Object {
      "Lines": false,
      "Words": true,
    }
  }
/>
`;
